gaia-header.js

```diff
@@ -34,105 +51,270 @@ module.exports = Component.register('gaia-header', {
    * @private
    */
   created: function() {
-    this.createShadowRoot().innerHTML = this.template;
+    debug('created');
+    this.setupShadowRoot();
 
-    // Get els
+    // Elements
     this.els = {
       actionButton: this.shadowRoot.querySelector('.action-button'),
-      headings: this.querySelectorAll('h1,h2,h3,h4'),
-      inner: this.shadowRoot.querySelector('.inner')
+      buttons: this.querySelectorAll('button, a'),
+      titles: this.querySelectorAll('h1')
     };
 
+    // Events
     this.els.actionButton.addEventListener('click', e => this.onActionButtonClick(e));
-    this.configureActionButton();
-    this.runFontFit();
+    this.observer = new MutationObserver(this.onMutation.bind(this));
+
+    // Properties
+    this.titleEnd = this.getAttribute('title-end');
+    this.titleStart = this.getAttribute('title-start');
+    this.noFontFit = this.getAttribute('no-font-fit');
+    this.notFlush = this.hasAttribute('not-flush');
+    this.action = this.getAttribute('action');
   },
 
   /**
    * Called when the element is
    * attached to the DOM.
    *
+   * Run fontFit when we have DOM
+   * context and start listening
+   * for DOM mutations.
+   *
    * @private
    */
   attached: function() {
-    this.rerunFontFit();
+    debug('attached');
+    this.nextTick(() => this.runFontFit());
+    this.observerStart();
   },
 
   /**
-   * Called when one of the attributes
-   * on the element changes.
+   * Called when the element is
+   * detached from the DOM.
    *
    * @private
    */
-  attributeChanged: function(attr) {
-    if (attr === 'action') {
-      this.configureActionButton();
-      this.rerunFontFit();
-    }
+  detached: function() {
+    debug('detached');
+    this.observerStop();
   },
 
   /**
-   * Runs the logic to size and position
-   * header text inside the available space.
+   * Run the font-fit logic and
+   * center the title.
    *
-   * @private
+   * Function calls ordering and use of
+   * `requestAnimationFrame` critical
+   * to minimising unwanted reflows.
+   *
+   * @public
    */
   runFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
-      fontFit.observeHeadingChanges(this.els.headings[i]);
-    }
+    debug('run font-fit');
+    if (this.noFontFit) { return; }
+    var space = this.getTitleSpace();
+
+    // Stop the mutation observer
+    // before making any DOM changes.
+    this.observerStop();
+
+    // Fit the text and then show each title.
+    // Once all titles have been fit, we can
+    // start observing DOM mutations again.
+    return Promise.all(map.call(this.els.titles, (el) => {
+      return this.fitTitleText(el, space);
+    })).then(() => { this.observerStart(); });
```

Comment:
there is a race problem here: what happens if a header user changes an attribute while we fit the text?

I think we should not stop the observer, but rather pause it, so that if it's called, we'd have a boolean "run again at resume". Of course we should log should it happen, because if this happens we should maybe add a nextTick somewhere, or wait longer.